marlin-succinct/protocol/src/prover.rs

```diff
@@ -0,0 +1,599 @@
+        let proof1: E::G1Affine;
+        let proof2: E::G1Affine;
+        let proof3: E::G1Affine;
+
+        // polynomial evaluations
+        let w_eval: E::Fr;
+        let za_eval: E::Fr;
+        let zb_eval: E::Fr;
+        let h1_eval: E::Fr;
+        let g1_eval: E::Fr;
+        let h2_eval: E::Fr;
+        let g2_eval: E::Fr;
+        let h3_eval: E::Fr;
+        let g3_eval: E::Fr;
+        let sigma2: E::Fr;
+        let sigma3: E::Fr;
+        let row_eval: [E::Fr; 3];
+        let col_eval: [E::Fr; 3];
+        let val_eval: [E::Fr; 3];
+
+        if index.compiled[0].constraints.shape().1 != witness.0.len() || witness.1 == 0
+        {
+            return Err(ProofError::WitnessCsInconsistent)
+        }
+
+        // random oracles have to be retrieved from the non-interactive argument
+        // context sequentually with adding argument-specific payload to the context
+
+        let mut oracles = RandomOracles::<E::Fr>::zero();
+
+        // prover computes z, w polynomials, second is the restriction of the first on the secret part of the witness
+        let z = Evaluations::<E::Fr>::from_vec_and_domain(witness.0.clone(), index.h_group).interpolate();
+
+        // extract secret of the witness
+        let mut w = witness.clone();
+        for i in 0..w.1 {w.0[i] = E::Fr::zero()}
+        let w = Evaluations::<E::Fr>::from_vec_and_domain(w.0, index.h_group).interpolate();
+
+        // prover computes za, zb polynomials
+        let mut zv = vec![vec![E::Fr::zero(); index.h_group.size()]; 2];
+
+        for i in 0..2
+        {
+            for constraint in index.compiled[i].constraints.iter()
+            {
+                zv[i][(constraint.1).0] += &(*constraint.0 * &witness.0[(constraint.1).1]);
+            }
+        }
+
+        // save public part of the witness
+        let mut public = witness.clone();
+        for i in public.1..public.0.len() {public.0[i] = E::Fr::zero()}
+
+        // prover interpolates the vectors and computes the evaluation polynomial
+        let za = Evaluations::<E::Fr>::from_vec_and_domain(zv[0].to_vec(), index.h_group).interpolate();
+        let zb = Evaluations::<E::Fr>::from_vec_and_domain(zv[1].to_vec(), index.h_group).interpolate();
+
+        // substitute ZC with ZA*ZB
+        let zv = [za.clone(), zb.clone(), &za * &zb]; 
+
+        // the transcript of the random oracle non-interactive argument
+        // the first part of the argument needs the following oracles:
+        // alpha, betta[0] in order to compute RA and evaluate Z, ZA & ZB polynomials
+
+        let mut argument = RandomOracleArgument::<E::Fr>::new(index.oracle_params.clone());
+        
+        argument.commit_scalar(&E::Fr::one()); // this commit comes from the previous proof context
+        oracles.alpha = argument.challenge();
+        // commit the public input into the argument
+        argument.commit_slice(&witness.0[0..witness.1]);
+        oracles.betta[0] = argument.challenge();
```

Comment:
Just a note on spelling, should be `beta`.

Reply:
Thank you for spotting...